src/main/java/net/dries007/tfc/common/recipes/BloomeryRecipe.java

```diff
@@ -0,0 +1,147 @@
+
+import net.dries007.tfc.common.recipes.ingredients.FluidStackIngredient;
+import net.dries007.tfc.common.recipes.ingredients.ItemStackIngredient;
+import net.dries007.tfc.common.recipes.inventory.ItemStackInventory;
+import net.dries007.tfc.util.JsonHelpers;
+
+public class BloomeryRecipe implements ISimpleRecipe<ItemStackInventory>
+{
+    private final ResourceLocation id;
+    private final FluidStackIngredient fluidStackIngredient;
+    private final ItemStackIngredient itemStackIngredient;
+    private final ItemStack result;
+    private final int time;
+
+    public BloomeryRecipe(ResourceLocation id, FluidStackIngredient fluidStackIngredient, ItemStackIngredient itemStackIngredient, ItemStack result, int time)
+    {
+        this.id = id;
+        this.fluidStackIngredient = fluidStackIngredient;
+        this.itemStackIngredient = itemStackIngredient;
+        this.result = result;
+        this.time = time;
+    }
+
+    public int getTime()
+    {
+        return time;
+    }
+
+    public ItemStackIngredient getCatalyst()
+    {
+        return itemStackIngredient;
+    }
+
+    public FluidStackIngredient getInputFluid()
+    {
+        return fluidStackIngredient;
+    }
+
+    @Override
+    public boolean matches(ItemStackInventory inv, Level level)
+    {
+        ItemStack stack = inv.getStack();
+        HeatingRecipe heatingRecipe = HeatingRecipe.getRecipe(stack);
+        FluidStack fluid = heatingRecipe == null ? FluidStack.EMPTY : heatingRecipe.getOutputFluid(new ItemStackInventory(stack));
+        return fluidStackIngredient.test(fluid) && itemStackIngredient.test(stack);
+    }
+
+    @Override
+    public ItemStack getResultItem()
+    {
+        return result;
+    }
+
+    public ItemStack assembleOutputs(FluidStack fluid)
+    {
+        return new ItemStack(this.result.getItem(),fluid.getAmount() / this.fluidStackIngredient.getAmount());
+    }
+
+    @Override
+    public ResourceLocation getId()
+    {
+        return id;
+    }
+
+    @Override
+    public RecipeSerializer<?> getSerializer()
+    {
+        return TFCRecipeSerializers.BLOOMERY.get();
+    }
+
+    @Override
+    public RecipeType<?> getType()
+    {
+        return TFCRecipeTypes.BLOOMERY.get();
+    }
+
+    public boolean isValidInput(ItemStack stack)
+    {
+        HeatingRecipe heatingRecipe = HeatingRecipe.getRecipe(stack);
+        if (heatingRecipe != null)
+        {
+            return this.fluidStackIngredient.test(heatingRecipe.getOutputFluid(new ItemStackInventory(stack)));
+        }
+        return false;
+    }
+
+    public boolean isValidMixture(FluidStack fluidStack, ItemStack catalystStack)
+    {
+        if (!fluidStackIngredient.test(fluidStack) || !itemStackIngredient.test(catalystStack))
+        {
+            return false;
+        }
+        return catalystStack.getCount() >= (fluidStack.getAmount() / fluidStackIngredient.getAmount() * itemStackIngredient.count());
+    }
```

Comment:
Neither of these methods should exist. `matches` is *exactly* what this method should be doing, and these are here because the recipe was designed with the wrong inventory choice.